torch/fx/graph.py

```diff
@@ -545,30 +635,82 @@ def python_code(self, root_module: str) -> str:
 
             The string source code generated from this ``Graph``.
         """
+        # Create a new nameespace for this Python source.
+        namespace = _Namespace()
+
+        # HACK: Override Node's repr to generate a valid name within our
+        # namespace. Since repr() is designed to produce a valid Python
+        # expression, it makes sense to re-use it. This way, it's easy to print
+        # something like Tuple[Node, Node] by simply calling repr() on it.
+        #
+        # Why can't we re-use node.name? Because it was generated within the
+        # namespace `self._graph_namespace`. In order to provide uniqueness
+        # guarantees within the Python code, we create a wholly new namespace
+        # to put all identifiers in.
+        def node_repr(self):
+            return namespace.create_name(self.name, self)
+
+        @contextmanager
```

Comment:
This is potentially a race condition. Instead I would just do the name lookup as part of a map_args in format args and return an object whose repr is itself:

```
def Self:
   def __init__(self, value):
     self.value = value
  def __repr__(self):
    return self.value

str(map_args(args, lambda node: Self(lookup_name_for_node(node)))
```

Reply:
this doesn't work that well because of how randomly nested things get (node's args can contain a node, or a list of nodes, or… etc.).

You're right that monkey patching the Node class is too dangerous though. I'll change it so that `Node.__repr__()` is cooperatively designed and we can just hook in a repr function on a per-instance basis.